### PR TITLE
[hail] promote `FilterIntervals` nodes to full IRs

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: hail/python/hail/ir/table_ir.py [hunk 1]
```diff
@@ -442,6 +442,24 @@ def _compute_type(self):
...
+    def _eq(self, other):
+        return self.intervals == other.intervals and self.point_type == other.point_type and self.keep == other.keep
+
+    def _compute_type(self):
+        self._type = self.child.typ
+
+
 class TableToTableApply(TableIR):
     def __init__(self, child, config):
         super().__init__(child)
```
Comment (danking) [inline hail/python/hail/ir/table_ir.py]: in `TableToTableApply._compute_type`, shouldn't you remove the if about `TableFilterIntervals`?

File: hail/python/test/hail/test_ir.py [hunk 2]
```diff
@@ -240,7 +241,9 @@ def matrix_irs(self):
             ir.MatrixExplodeCols(matrix_read, ['col_aset']),
             ir.MatrixAnnotateRowsTable(matrix_read, table_read, '__foo'),
             ir.MatrixAnnotateColsTable(matrix_read, table_read, '__foo'),
-            ir.MatrixToMatrixApply(matrix_read, {'name': 'MatrixFilterPartitions', 'parts': [0], 'keep': True})
+            ir.MatrixToMatrixApply(matrix_read, {'name': 'MatrixFilterPartitions', 'parts': [0], 'keep': True}),
+            ir.MatrixRename(matrix_read, {'global_f32': 'global_foo'}, {'col_f32': 'col_foo'}, {'row_aset': 'row_aset2'}, {'entry_f32': 'entry_foo'}),
```
Comment (danking) [inline hail/python/test/hail/test_ir.py]: why the rename test?

Comment (tpoterba) [reply hail/python/test/hail/test_ir.py]: was missing.

File: hail/python/hail/ir/table_ir.py (hunk 1 above)
Comment (tpoterba) [reply hail/python/hail/ir/table_ir.py]: yes!

File: hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 3]
```diff
@@ -505,7 +505,12 @@ object PruneDeadFields {
           globalType = requestedType.globalType.rename(globalMapRev),
           key = requestedType.key.map(k => rowMapRev.getOrElse(k, k)))
         memoizeTableIR(child, childDep, memo)
-      case TableToTableApply(child, f) => memoizeTableIR(child, f.requestType(requestedType, child.typ), memo)
+      case TableFilterIntervals(child, _, _) =>
+        memoizeTableIR(child, requestedType.copy(key = child.typ.key,
+          rowType = PruneDeadFields.unify(child.typ.rowType,
+            requestedType.rowType,
+            PruneDeadFields.selectKey(child.typ.rowType, child.typ.key))), memo)
+      case TableToTableApply(child, f) => memoizeTableIR(child, child.typ, memo)
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: I added a requested type mechanic to some of the relational functions when I first added the interval pushdown stuff; this was a hack. The right solution is making FilterIntervals IRs full nodes, as in this PR

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala [hunk 4]
```diff
@@ -530,6 +530,38 @@ object Simplify {
...
+        Interval.intersection(i1.toArray[Interval], i2.toArray[Interval], ord)
+      else
+      // remove means union intervals
+        Interval.union(i1.toArray[Interval] ++ i2.toArray[Interval], ord)
+      TableFilterIntervals(child, intervals.toFastIndexedSeq, k1)
+    // TODO: write rule for TableParallelize that pushes down filter into value IR
+    // TODO: write rule for TableKeyByAndAggregate that pushes down filter as a TableFilter
+    // TODO: write rule for TableRead that that pushes down filter into coming index files
+    // TODO: write rule for TableJoin that pushes down filter to both sides
+    // TODO: write rule for TableKeyBy(TableFilterIntervals(TableKeyBy))), which could prevent key pruning optimization
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: good plan, done

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala [hunk 5]
```diff
@@ -530,6 +530,38 @@ object Simplify {
...
+    case TableFilterIntervals(TableIntervalJoin(child, right, root, product), intervals, keep) =>
+      TableIntervalJoin(TableFilterIntervals(child, intervals, keep), right, root, product)
+    case TableFilterIntervals(TableExplode(child, path), intervals, keep) =>
+      TableExplode(TableFilterIntervals(child, intervals, keep), path)
+    case TableFilterIntervals(TableAggregateByKey(child, expr), intervals, keep) =>
+      TableAggregateByKey(TableFilterIntervals(child, intervals, keep), expr)
+    case TableFilterIntervals(TableFilterIntervals(child, i1, k1), i2, k2) if k1 == k2 =>
+      val ord = child.typ.keyType.ordering.intervalEndpointOrdering
+      val intervals = if (k1)
+      // keep means intersect intervals
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: done

**Pull request merged**